File: code/matrices.go
package main

import (
	"fmt"
	"math"
	"time"
)

func multMatrizClasica(A, B, res [][]int, N int) {
	for i := 0; i < N; i++ {
		for j := 0; j < N; j++ {
			for k := 0; k < N; k++ {
				res[i][j] += A[i][k] * B[k][j]
			}
		}
	}
}

func multMatrizBloque(A, B, res [][]int, N, S int) {
	for ii := 0; ii < N; ii += S {
		for jj := 0; jj < N; jj += S {
			for kk := 0; kk < N; kk += S {
				for i := ii; i < int(math.Min(float64(ii+S), float64(N))); i++ {
					for j := jj; j < int(math.Min(float64(jj+S), float64(N))); j++ {
						for k := kk; k < int(math.Min(float64(kk+S), float64(N))); k++ {
							res[i][j] += A[i][k] * B[k][j]
						}
					}
				}
			}
		}
	}
}


func createMatrix(N int) [][]int {
	matrix := make([][]int, N)
	for i := range matrix {
		matrix[i] = make([]int, N)
	}
	return matrix
}

func main() {
	sizes := []int{10, 20, 40, 80, 160, 320, 640, 1280, 2560}

	fmt.Println("Tamaño matriz\tMultClasica\tMultBloque")

	for _, size := range sizes {
		A := createMatrix(size)
		B := createMatrix(size)
		res := createMatrix(size)

		
		value := 1
		for i := 0; i < size; i++ {
			for j := 0; j < size; j++ {
				A[i][j] = value
				B[i][j] = value + 1
				value++
			}
		}

		
		start1 := time.Now()
		multMatrizClasica(A, B, res, size)
		elapsed1 := time.Since(start1)

		
		S := 80 
		start2 := time.Now()
		multMatrizBloque(A, B, res, size, S)
		elapsed2 := time.Since(start2)

		fmt.Printf("%d\t%.6f\t%.6f\n", size, elapsed1.Seconds(), elapsed2.Seconds())
	}
}

File: code/matrices.cpp

#include <iostream>
#include <chrono>
#include <algorithm>
using namespace std;

void multMatrizClasica(int **m1, int **m2, int **res, int N ){
   
    for(int i = 0; i < N;i++){
        for(int j = 0; j < N; j++){
            for(int k = 0; k < N; k++){
                *(*(res+i)+j) += *(*(m1+i)+k) * *(*(m2+k)+j);
            }
        }
    }
    return;
}


void multMatrizBloque(int **m1, int **m2, int **res, int N , int S){
    for (int ii = 0; ii < N; ii += S) {
        for (int jj = 0;jj < N; jj += S) {
            for (int kk = 0; kk < N; kk += S) {
                for (int i = ii;i < std::min(ii+S,N); ++i) {
                    for (int j = jj;j < std::min(jj+S, N); ++j) {
                        for (int k = kk; k < std::min(kk +S, N);++k) {
                            *(*(res+i)+j) += *(*(m1+i)+k) * *(*(m2+k)+j);

                            //res[i][j] += m1[i][k]*m2[k][j];
                        }
                    }
                }
            }
        }
    }
}
int main() {
    int sizes[] = {10, 20, 40, 80, 160, 320, 640, 1280, 2560};
    int blockSizes[] = {32, 64, 128, 256}; 

    cout << "Tiempo en segundos\n";
    cout << "Tamaño Matriz\tTamaño Bloque\tMultClásica\tMultBloque\n";

    for (int size : sizes) {
        
        int **A = new int*[size];
        int **B = new int*[size];
        int **res = new int*[size];

        for (int i = 0; i < size; ++i) {
            A[i] = new int[size];
            B[i] = new int[size];
            res[i] = new int[size];
        }

        int value = 1;
        for (int i = 0; i < size; ++i) {
            for (int j = 0; j < size; ++j) {
                A[i][j] = value++;
                B[i][j] = value++;
                res[i][j] = 0;
            }
        }

        auto start1 = chrono::high_resolution_clock::now();
        multMatrizClasica(A, B, res, size);
        auto end1 = chrono::high_resolution_clock::now();
        chrono::duration<double> duration1 = end1 - start1;

        for (int S : blockSizes) {
            for (int i = 0; i < size; ++i)
                fill(res[i], res[i] + size, 0);
            auto start2 = chrono::high_resolution_clock::now();
            multMatrizBloque(A, B, res, size, S);
            auto end2 = chrono::high_resolution_clock::now();
            chrono::duration<double> duration2 = end2 - start2;

            cout << size << "\t\t" << S << "\t\t" 
                 << duration1.count() << "\t\t" 
                 << duration2.count() << endl;
        }

        for (int i = 0; i < size; ++i) {
            delete[] A[i];
            delete[] B[i];
            delete[] res[i];
        }
        delete[] A;
        delete[] B;
        delete[] res;
    }

    return 0;
}
